Extract Container class from Engine

pbrisbin opened the pull request:
New responsibility breakdown:

**Container**: launch a docker container, give its stdout to the caller,
capture its stderr, monitor its process, handle timeouts, call life-cycle
events on the container-log given.

**Engine::ContainerLog**: perform the statsd calls on container life-cycle
events and then forward those events to the log its been given; sort of a
ContainerLog middleware stack idea.

**Engine**: launch and run a Container for the engine, specify how to shuttle
null-delimited stdout to the formatter, compose its container-log with the one
given.

Benefits:

- Engine is smaller
- Engine#run is smaller
- Container is singly-responsible and reusable for Clone in builder
- Easier to reason about (complex timeout logic is clearer IMO)
- Better tested

Notable change:

Rather than spawning `docker kill` I'm using Ruby's `Process.kill`, the two
should be equivalent and this was more convenient with the way the code is now
(no need to track container-name).

GordonDiggs commented on lib/cc/analyzer/container.rb:
```diff
@@ -0,0 +1,98 @@
+        log: NullContainerLog.new,
+        timeout: DEFAULT_TIMEOUT
+      )
+        @image = image
```
Is there a reason not to use `attr_reader` for these?

pbrisbin replied:
I don't have a strong preference between making (private) `attr_reader`s or using instance variables directly. I think I advocated the latter at one point but because I see the former more prevalent our code bases I've been starting to do that -- which is what I've done here.

GordonDiggs replied:
>  I see the former more prevalent our code bases I've been starting to do that -- which is what I've done here.

I don't see `attr_reader` in this class

pbrisbin replied:
Sorry, got my former and latters mixed up. Editing my original comment to hopefully clarify...

GordonDiggs replied:
Cool. I haven't seen much of using instance variables directly (probably because I prefer the `attr_reader` approach).

Functionally, they're the same so I don't see a big win either way.

pbrisbin replied:
> I haven't seen much of using instance variables directly

I'd have to do some kind of audit to be sure, but at least in this PR `Engine` also accesses its instance variables directly.

I was going to switch these over, but it's actually kind of annoying and error prone so I'm going to leave them as is (if that's OK with you). If you want to push for the `attr_reader` approach can you open a PR on the style guide?

GordonDiggs replied:
Will do!

GordonDiggs commented on lib/cc/analyzer/engine/container_log.rb:
```diff
@@ -0,0 +1,48 @@
+      class ContainerLog
+        def initialize(name, inner_log)
+          @name = name
+          @inner_log = inner_log
```
same `attr_reader` question here

ABaldwinHunter commented on lib/cc/analyzer/container.rb:
```diff
@@ -0,0 +1,98 @@
+      DEFAULT_TIMEOUT = 15 * 60 # 15m
+
+      def initialize(
+        image:,
```
@pbrisbin this is probably a ruby syntax I'm not familiar with, but just wanted to make sure the absence of a value here was intentional.

pbrisbin replied:
Yup, makes it a required keyword argument.

ABaldwinHunter replied:
Cool. thanks for explanation!

GordonDiggs commented:
LGTM. 2 questions:

1. Do any of the new issues on CC need to be addressed?
2. Does this work locally?

pbrisbin commented:
> Do any of the new issues on CC need to be addressed?

The issues on CC are because the `#run` method which I moved from `Engine` to `Container`, while reduced and simplified greatly (IMO), is still flagged -- (hence 5 fixed in Engine and 5 new in Container). IMO the class/method is improved and there's not much obvious work I can do to go further such that CC is also happy so I'm good shipping it as is. WDYT?

> Does this work locally?

Will definitely test locally before shipping.

GordonDiggs commented:
I'm cool with that. Should we disable the `Space inside empty braces detected.` rule?

pbrisbin commented:
I'm not sure what that'd do. I do want to catch anyone writing `{ }` for a Hash or `[ ]` for an Array, but a I don't want to write `{}` for a lambda...

pbrisbin commented:
FYI: `codeclimate analyze` works locally
